### PR TITLE
Feat/constructor

### PR DESCRIPTION
**Pull request by internnos**
this is the implementation of `constructor` parser

TODO:
  - ~running parser on the whole scope~
  - ~output the comment parsing result into a proper data structure~
  - ~refactor (to new class) if necessary~
  - ~create baseclass if necessary~

### CODE REVIEW

**Comment by AnthonyBuisset** [comment]
Nice work so far !

A couple of improvements ideas:

It seems your parser as too many responsibilities:
   - Finding a corresponding function in the file
   - parsing the function signature
   - parsing the function comments

I think it would be wise to split into different modules (check [SOLID principles](https://en.wikipedia.org/wiki/SOLID))
Then you can create an main entry class that will interact with those above.

For the function signature parsing, another solution would be to rely on a starknet/cairo compiler (cairo-compile, starknet-compile, protostar, nile).
Pros:
   - your code won't be affected if the language evolves
   - the output is a structured json file with all data you need (check example for instance [basic_ship.json.zip](https://github.com/onlydustxyz/kaaper/files/9027371/basic_ship.json.zip), under the identifiers section)
Cons:
   - you have to integrate a python module and run it within your code
   - the parsing of the code depends on the compiler, so might take longer

The good thing if you split into modules as mentionned above, you will be able to change the implementation later, or have it configurable :-)

**Comment by internnos** [comment]
> Nice work so far !
> 
> A couple of improvements ideas:
> 
> It seems your parser as too many responsibilities:
> 
> * Finding a corresponding function in the file
> * parsing the function signature
> * parsing the function comments
> 
> I think it would be wise to split into different modules (check [SOLID principles](https://en.wikipedia.org/wiki/SOLID)) Then you can create an main entry class that will interact with those above.
> 

i really love this idea! will re-structure it in this way

> For the function signature parsing, another solution would be to rely on a starknet/cairo compiler (cairo-compile, starknet-compile, protostar, nile). Pros:
> 
> * your code won't be affected if the language evolves
> * the output is a structured json file with all data you need (check example for instance [basic_ship.json.zip](https://github.com/onlydustxyz/kaaper/files/9027371/basic_ship.json.zip), under the identifiers section)
>   Cons:
> * you have to integrate a python module and run it within your code
> * the parsing of the code depends on the compiler, so might take longer
> 

good idea! tbh i've thought the exact the same thing but couldn't came up with good implementation so i sticked with the pure regex instead. anyway is there any compiler that run on js? would be a game changer if there's one


> The good thing if you split into modules as mentionned above, you will be able to change the implementation later, or have it configurable :-)

i highly agree with this. thank you!

**Comment by internnos** [comment]
> > is there any compiler that run on js? would be a game changer if there's one
> 
> Not that I know of, only python

alrite, but it's still good idea. i think ill make different version which uses compiler, so when there's breaking change in the syntax, `kaaper` is still usable. thanks for the suggestion!

File: src/lib/parser/function-comment/implicit-args.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import { ServerResponse } from "http";
```
Comment (AnthonyBuisset) [inline src/lib/parser/function-comment/implicit-args.ts:1]: why is this needed ? spoiler alert ? :-)

Comment (internnos) [reply src/lib/parser/function-comment/implicit-args.ts:1]: aah sorry i think let this slide during the paste process

will remove!

File: src/test/suite/parser.test.ts [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+  test("get desc comments", () => {
+    const pathFile = path.resolve(
+      __dirname,
+      "../../../test_assets/ERC20.cairo"
+    );
+    let constructorParser = new CairoParser(pathFile, "constructor");
+    const attributeScope = constructorParser.parseAttribute();
+    const comments = constructorParser.getComments(attributeScope);
+    if (comments) {
```
[the remark this answers is not in the document]
Comment (internnos) [reply src/test/suite/parser.test.ts]: aah gotcha

make sense

will do!

File: src/lib/parser/function-comment/explicit-args.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
+export default class FunctionCommentExplicitArgsParser {
```
[the remark this answers is not in the document]
Comment (internnos) [reply src/lib/parser/function-comment/explicit-args.ts]: i've updated this

any comments/suggestions?

thank you

**Review by AnthonyBuisset — APPROVED** [review]

**Review by AnthonyBuisset — APPROVED** [review]
